src/rules.mli

```diff
@@ -0,0 +1,21 @@
+(** [Rules] represents a collection of rules across a known set of directories. *)
+
+open! Stdune
+
+(** [rule] is a function that produces some build system rules
+    such as ([Build_system.add_rule]) in a known directory. *)
+type rule = unit -> unit
+
+type t = private rule Path.Build.Map.t
+
+val to_map : t -> rule Path.Build.Map.t
+
+(* [Path] must be in build directory *)
+val file_rule : rule:(Path.t * rule) -> unit
```

Comment:
Should we use `Path.Build.t` here?

Reply:
Yes, as soon as `build_system.ml` is switched (this will be done in a follow-up feature).

Reply:
In fact this occurrence of `Path.t` is just removed later (because the `Path.t` information is duplicated in the rule itself):

```
https://github.com/ocaml/dune/pull/2089/files#diff-9cb90475c596eb47885cc09b7f1af8ce
```

Reply:
Ack